pqarrow/arrow.go

```diff
@@ -82,10 +73,142 @@ func rowBasedParquetRowGroupToArrowRecord(
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+				name := field.Name()
+				if distinctColumns[0].Match(name) {
+					typ, nullable, _, err := parquetColumnToArrowArray(
+						pool,
+						field,
+						parquetColumns[i],
+						true,
+					)
+					if err != nil {
+						return nil, fmt.Errorf("convert parquet column to arrow array: %w", err)
+					}
+					fields = append(fields, arrow.Field{
+						Name:     name,
+						Type:     typ,
+						Nullable: nullable,
+					})
+				}
+			}
+		}
+
+		return arrow.NewSchema(fields, nil), nil
+	}
+
+	fields := make([]arrow.Field, 0, len(parquetFields))
+
+	for i, parquetField := range parquetFields {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+			if includedProjection(projections, parquetField.Name()) {
+				typ, nullable, _, err := parquetColumnToArrowArray(
+					pool,
+					parquetField,
+					parquetColumns[i],
+					false,
+				)
+				if err != nil {
+					return nil, fmt.Errorf("convert parquet column to arrow array: %w", err)
+				}
+				fields = append(fields, arrow.Field{
+					Name:     parquetField.Name(),
+					Type:     typ,
+					Nullable: nullable,
+				})
+			}
+		}
+	}
+
+	return arrow.NewSchema(fields, nil), nil
+}
+
+func includedProjection(projections []logicalplan.ColumnMatcher, name string) bool {
+	if len(projections) == 0 {
+		return true
+	}
+
+	for _, p := range projections {
+		if p.Match(name) {
+			return true
+		}
+	}
+	return false
+}
+
+// ParquetRowGroupToArrowRecord converts a parquet row group to an arrow record.
+func ParquetRowGroupToArrowRecord(
+	ctx context.Context,
+	pool memory.Allocator,
+	rg parquet.RowGroup,
+	schema *arrow.Schema, // TODO!
+	filterExpr logicalplan.Expr,
+	distinctColumns []logicalplan.ColumnMatcher,
+) (arrow.Record, error) {
+	switch rg.(type) {
+	case *dynparquet.MergedRowGroup:
+		return rowBasedParquetRowGroupToArrowRecord(ctx, pool, rg, schema)
+	default:
+		return contiguousParquetRowGroupToArrowRecord(
+			ctx,
+			pool,
+			rg,
+			schema,
+			filterExpr,
+			distinctColumns,
+		)
+	}
+}
+
+// rowBasedParquetRowGroupToArrowRecord converts a parquet row group to an arrow record row by row.
+func rowBasedParquetRowGroupToArrowRecord(
+	ctx context.Context,
+	pool memory.Allocator,
+	rg parquet.RowGroup,
+	schema *arrow.Schema,
+) (arrow.Record, error) {
+	parquetFields := rg.Schema().Fields()
+
+	if len(schema.Fields()) != len(parquetFields) {
+		return nil, fmt.Errorf("inconsistent schema between arrow and parquet")
+	}
+
+	// Create arrow writers from arrow and parquet schema
 	writers := make([]writer.ValueWriter, len(parquetFields))
-	b := array.NewRecordBuilder(pool, arrow.NewSchema(fields, nil))
+	b := array.NewRecordBuilder(pool, schema)
 	for i, column := range b.Fields() {
-		writers[i] = newWriterFuncs[i](column, 0)
+		parquetField := parquetFields[i]
+		_, newValueWriter, err := convert.ParquetNodeToTypeWithWriterFunc(parquetField)
```

Comment:
It feels like we should have a function other than `ParquetNodeToTypeWithWriterFunc` now, that only returns what we truly need

Reply:
Looking at the rest of the file only here we don't use the returned datatype. Everywhere else we do, so I'm fine having it stay like that.